### PR TITLE
Simplify Storage meta conversion with PyObject preservation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122044
* __->__ #122018

Thanks to https://github.com/pytorch/pytorch/pull/109039 we can rely on
finalizers on Storage PyObject to handle removal from dict.

Irritatingly, we still have to attach finalizer, because we don't have
a weak key AND value dict (only one or the other).

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -165,36 +137,42 @@ def del_ten():
...
+        return self.storage_memo.get(WeakIdRef(s), None)
+
+    def set_storage_memo(self, s, v):
+        # hold a weak ref to self, otherwise it will be kept alive
+        # by the del_ten closure
+        self_weak_ref = weakref.ref(self)
+        storage_ref_key = WeakIdRef(s)
+
+        def del_storage():
+            # tensor outlives the converter
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:154]: copy-pasta `tensor`

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -165,36 +137,42 @@ def del_ten():
...
+    def get_storage_memo(self, s):
+        return self.storage_memo.get(WeakIdRef(s), None)
+
+    def set_storage_memo(self, s, v):
+        # hold a weak ref to self, otherwise it will be kept alive
+        # by the del_ten closure
+        self_weak_ref = weakref.ref(self)
+        storage_ref_key = WeakIdRef(s)
+
+        def del_storage():
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:153]: should this be factored out with `del_tensor` ?

**Review by eellison — APPROVED** [review]
nice!

**Review by kurtamohler — APPROVED** [review]

File: torch/_subclasses/meta_utils.py (hunk 2 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:153]: This gets simplified EVEN MORE in the next PR, so just leaving as is.

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:154]: This gets deleted in next PR, leaving it there.

**Pull request closed without merging**